File: app/src/main/java/com/techmave/fisherville/dialog/OtpDialog.java
package com.techmave.fisherville.dialog;

import android.graphics.Point;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.text.TextUtils;
import android.view.Display;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatEditText;
import androidx.fragment.app.DialogFragment;

import com.techmave.fisherville.R;
import com.techmave.fisherville.callback.OnOtpReceived;

import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class OtpDialog extends DialogFragment {

    @BindView(R.id.otp_text)
    AppCompatEditText otpText;
    @BindView(R.id.otp_resend)
    AppCompatButton otpResend;

    private OnOtpReceived listener;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.dialog_otp, container, false);
        ButterKnife.bind(this, view);

        countDown();

        return view;
    }

    private void countDown() {

        String text = Objects.requireNonNull(getContext()).getResources().getString(R.string.resend);

        CountDownTimer timer = new CountDownTimer(300000, 1000) {

            @Override
            public void onTick(long millisUntilFinished) {

                long seconds = millisUntilFinished / 1000;
                long minutes;
                String temp = text;

                if (seconds >= 60) {

                    minutes = seconds / 60;
                    seconds = seconds - minutes * 60;

                    if (minutes > 9) {

                        if (seconds > 9) {

                            temp += " (" + minutes + ":" + seconds + ")";

                        } else {

                            temp += " (" + minutes + ":0" + seconds + ")";
                        }

                    } else {

                        if (seconds > 9) {

                            temp += " (0" + minutes + ":" + seconds + ")";

                        } else {

                            temp += " (0" + minutes + ":0" + seconds + ")";
                        }
                    }

                } else {

                    if (seconds > 9) {

                        temp += " (00:" + seconds + ")";

                    } else {

                        temp += " (00:0" + seconds + ")";
                    }
                }

                otpResend.setText(temp);
            }

            @Override
            public void onFinish() {

                otpResend.setEnabled(true);
            }
        };

        timer.start();
    }

    @Override
    public void onResume() {

        super.onResume();

        Window window = Objects.requireNonNull(getDialog()).getWindow();
        Point size = new Point();

        if (window != null) {

            Display display = window.getWindowManager().getDefaultDisplay();
            display.getSize(size);

            int width = size.x;

            window.setLayout((int) (width * 0.90), WindowManager.LayoutParams.WRAP_CONTENT);
            window.setGravity(Gravity.CENTER_HORIZONTAL);
        }
    }

    public void setOnOtpReceived(OnOtpReceived listener) {

        this.listener = listener;
    }

    @OnClick({R.id.otp_submit, R.id.otp_resend})
    public void onViewClicked(View view) {

        switch (view.getId()) {

            case R.id.otp_submit:
                retrieveCode();
                break;

            case R.id.otp_resend:
                break;
        }
    }

    private void retrieveCode() {

        if (otpText.getText() == null || TextUtils.isEmpty(otpText.getText().toString())) {

            otpText.setError("Required");

        } else if (otpText.getText().toString().length() != 6) {

            otpText.setError("Incorrect OTP");

        } else {

            listener.onReceived(otpText.getText().toString());
            dismiss();
        }
    }
}

File: app/src/main/java/com/techmave/fisherville/view/dialog/OtpDialog.kt
package com.techmave.fisherville.view.dialog

import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.Gravity
import android.view.View
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.techmave.fisherville.R
import com.techmave.fisherville.databinding.DialogOtpBinding
import com.techmave.fisherville.util.extension.LifeCycleExtensions.observeOnce
import com.techmave.fisherville.util.extension.LifeCycleExtensions.setNavigationResult
import com.techmave.fisherville.viewmodel.LoginViewModel
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class OtpDialog @Inject constructor(): BaseDialog<DialogOtpBinding>(DialogOtpBinding::inflate) {

    private val viewModel: LoginViewModel by viewModels()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        super.onViewCreated(view, savedInstanceState).also { setWidthPercentage(90) }
    }

    override fun onResume() {

        super.onResume().also {

            dialog?.window?.apply {

                setGravity(Gravity.CENTER_HORIZONTAL)
                setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            }
        }
    }

    override fun initialize() {

        binding.viewModel = viewModel
        binding.lifecycleOwner = viewLifecycleOwner

        binding.otpCancel.setOnClickListener { findNavController().popBackStack() }

        viewModel.otpValidated.observe(viewLifecycleOwner) {

            if (it) {

                viewModel.mainOtp.observeOnce(viewLifecycleOwner) { otp ->

                    setNavigationResult(resources.getString(R.string.action_otp), otp).also { findNavController().navigateUp() }
                }
            }
        }
    }

//    private fun countDown() {
//
//        val text = context?.getString(R.string.resend)
//        val timer: CountDownTimer = object : CountDownTimer(300000, 1000) {
//
//            override fun onTick(millisUntilFinished: Long) {
//
//                var seconds = millisUntilFinished / 1000
//                val minutes: Long
//                var temp = text
//
//                if (seconds >= 60) {
//
//                    minutes = seconds / 60
//                    seconds -= minutes * 60
//                    temp += if (minutes > 9) {
//
//                        if (seconds > 9) " ($minutes:$seconds)" else " ($minutes:0$seconds)"
//
//                    } else {
//
//                        if (seconds > 9) " (0$minutes:$seconds)" else " (0$minutes:0$seconds)"
//                    }
//
//                } else {
//
//                    temp += if (seconds > 9) " (00:$seconds)" else " (00:0$seconds)"
//                }
//
//                binding.otpResend.text = temp
//            }
//
//            override fun onFinish() {
//
//                binding.otpResend.isEnabled = true
//            }
//        }
//
//        timer.start()
//    }

//    override fun onClick(v: View?) {
//
//        when (v?.id) {
//
//            R.id.otp_submit -> validateCode()
//        }
//    }
//
//    private fun validateCode() {
//
//        when {
//
//            binding.otpText.text.isNullOrEmpty() -> binding.otpText.error = "Required"
//            binding.otpText.text.toString().length != 6 -> binding.otpText.error = "Incorrect Code Length"
//
//            else -> {
//
//                listener?.onOtpReceived(binding.otpText.text.toString())
//                dismiss()
//            }
//        }
//    }
}